components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -449,6 +473,39 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
 			"timeout": "300s",
 		},
 	}
+
+	// Create a template map for the headers section
+	// of the Istio VirtualService
+	headers := map[string]interface{}{
+		"request": map[string]interface{}{
+			"set": map[string]interface{}{},
+		},
+	}
+
+	// Blacklist for request headers that are not allowed to be set
+	blacklist := []string{"userid-header", "X-Forwarded-Proto"}
+
+	// If key "notebooks.kubeflow.org/http-headers-request-set" is present on the notebook resource,
+	// split the literal string into keys and values and add them
+	// to the headers variable. Finally, add the headers section to
+	// the variable with the http spec.
+	if _, ok := annotations["notebooks.kubeflow.org/http-headers-request-set"]; ok && len(annotations["notebooks.kubeflow.org/http-headers-request-set"]) > 0 {
+		requestHeaders := strings.Split(annotations["notebooks.kubeflow.org/http-headers-request-set"], "\n")
+		for _, kv := range requestHeaders {
+			if len(strings.Split(kv, ": ")) == 2 {
+				k := strings.Split(kv, ": ")[0]
+				v := strings.Split(kv, ": ")[1]
```

Comment:
I believe we shouldn't use our own serialization format, but instead take a page from Kubernetes' book.
Kubernetes uses annotations for alpha features. How are those annotations written? With JSON. For example, the `scheduler.alpha.kubernetes.io/affinity` annotation.

Using JSON allows us to leverage existing battle-tested parsers and not implement our own. What do you think?

Reply:
I have no idea how to start implementing this, and I don't believe there is enough time to do so anymore either. I'm mainly following what is already being done in the code elsewhere. If you want to improve the code, I'm happy to include it in the PR.

Reply:
@DavidSpek it's a small change. Here is an example you can follow to unmarshal a json map (the use-case we have here):
https://play.golang.org/p/TR-6kdk6s91

Reply:
@DavidSpek show your comment below about the error with Unmarshal https://github.com/kubeflow/kubeflow/pull/5660#issuecomment-797378906. Writing it here to keep it in one place.

I'm not sure how you used the Unmarshal function but I believe your could also try and catch the error and not let it panic. Here's a very similar code I've seen https://github.com/kubeflow/katib/blob/master/pkg/new-ui/v1beta1/backend.go#L73

Reply:
@DavidSpek in the example I gave you, I omitted the error handling. Let me give you an example with error handling. This way you can catch the bad case in the controller, log it and emit an event:
https://play.golang.org/p/nxmAh3VaNZr

Reply:
@yanniszark Thanks for the example. I just noticed the actual latest changes were not pushed yesterday. I just pushed them and I believe the error handling from you example is included in there. However, the controller crashing is still a problem.

Reply:
Then we need to find the cause. Can you point to the exact line in your code that crashes?
What have you tried / what do you think is wrong?